Initial implementation

hosamaly opened the pull request:
The best way to read this PR, in my opinion, is to ignore the generated files in the 3.3, 3.4, 3.5, and 3.6 directories in the beginning. I'd strongly suggest starting with the README first.

hosamaly committed:
Create .gitignore configured for Scala and SBT

hosamaly committed:
Add the NITF XML Schema Definition files

hosamaly committed:
Create sbt build infrastructure

The infrastructure is configured to generate projects dynamically.
This makes it easier to add or remove supported versions as desired.

hosamaly committed:
Add generated Scala files as well as tests

Although they are generated, I think that it's important to be able to
track their changes, so I'm adding them to the git repository.

hosamaly committed:
Add configuration for Travis CI

LATaylor-guardian commented on project/Build.scala:
```diff
@@ -0,0 +1,44 @@
+  lazy val testProjects: Seq[Project] = schemaVersions.map(testProject)
+
+  private val releaseVersion = 0
+  private val schemaVersions = Seq(/*"3.3",*/ "3.4", "3.5", "3.6")
```
Why is 3.3 commented out? Can this be deleted?

hosamaly replied:
It shouldn't be. I did a `git reset` and forgot some of the changes. Thanks for noticing!

LATaylor-guardian commented on project/Build.scala:
```diff
@@ -0,0 +1,44 @@
+  )
+
+  private lazy val testSettings = Seq(
+//    scalaVersion := Dependencies.scala211Version,  // scala-xml-diff is released for 2.11 only
```
And here?

LATaylor-guardian commented on project/Build.scala:
```diff
@@ -0,0 +1,44 @@
+
+import net.vonbuchholtz.sbt.dependencycheck.DependencyCheckPlugin.autoImport.dependencyCheckFailBuildOnCVSS
+
+object Build extends BuildDef with BuildCommon {
```
Is there any reason you are using a `Build.scala` file? `Build.scala` is deprecated in preference of a simple `build.sbt` file.

hosamaly replied:
It's because I wanted to generate the projects dynamically. Using `build.sbt`, I was only able to define them manually, one by one.

hosamaly committed:
Uncomment v3.3

It was commented while debugging an issue that no longer exists.

LATaylor-guardian commented on project/Dependencies.scala:
```diff
@@ -0,0 +1,24 @@
+import sbt._
+
+object Dependencies {
```
I'm not sure you gain much by splitting dependencies out of build.sbt at this point? The project (and it's configuration) is small enough to reside in a single `build.sbt` file.

hosamaly replied:
The sbt documentation (to which I'm new) said that it's a best practice to split dependencies into their own class. I don't mind merging it into `Build.scala` or `build.sbt`, although it did feel a bit cleaner to separate them.

hosamaly committed:
Speed build up on Travis CI by running parallel jobs for each version

regiskuckaertz commented on project/Build.scala:
```diff
@@ -0,0 +1,43 @@
+  lazy val mainProjects: Seq[Project] = schemaVersions.map(nitfProject)
+  lazy val testProjects: Seq[Project] = schemaVersions.map(testProject)
+
+  private val releaseVersion = 0
```
Do we have to update that value on each new release? Is there a way to move that out into a `version.sbt` file?

hosamaly replied:
I was thinking that we'll get rid of this by adding the `sbt-release` plugin and configuring it to manage this automatically. I'm _guessing_ that the plugin would generate `version.sbt` automatically. Is this correct?

regiskuckaertz replied:
Yes indeed

hosamaly committed:
Improve caching in Travis CI

Following advice from:
https://www.scala-sbt.org/1.x-beta/docs/Travis-CI-with-sbt.html

hosamaly committed:
Test against examples specific for each version of NITF

Some of the examples are exact copies of others.
Some differ only by the schema version declared at the top of the file.
Some are totally different from the others.

tomrf1 commented:
👍 looks good
